packages/next/src/client/components/bfcache.ts

```diff
@@ -0,0 +1,122 @@
+  // without cloning, unless there's a reordering or removal.
+  // TODO: Once we start tracking back/forward history at each route level,
+  // we should use the history order instead. In other words, when traversing
+  // to an existing entry as a result of a popstate event, we should maintain
+  // the existing order instead of moving it to the front of the list. I think
+  // an initial implementation of this could be to pass an incrementing id
+  // to history.pushState/replaceState, then use that here for ordering.
+  const [prevActiveEntry, setPrevActiveEntry] = useState<RouterBFCacheEntry>(
+    () => {
+      const initialEntry: RouterBFCacheEntry = {
+        tree: activeTree,
+        stateKey: activeStateKey,
+        next: null,
+      }
+      return initialEntry
+    }
+  )
+
+  if (prevActiveEntry.tree === activeTree) {
+    // Fast path. The active tree hasn't changed, so we can reuse the
+    // existing state.
+    return prevActiveEntry
+  }
+
+  // The route tree changed. Note that this doesn't mean that the tree changed
+  // *at this level* — the change may be due to a child route. Either way, we
+  // need to either add or update the router tree in the bfcache.
+  //
+  // The rest of the code looks more complicated than it actually is because we
+  // can't mutate the state in place; we have to copy-on-write.
+
+  // Create a new entry for the active cache key. This is the head of the new
+  // linked list.
+  const newActiveEntry: RouterBFCacheEntry = {
+    tree: activeTree,
+    stateKey: activeStateKey,
+    next: null,
+  }
+
+  // We need to append the old list onto the new list. If the head of the new
+  // list was already present in the cache, then we'll need to clone everything
+  // that came before it. Then we can reuse the rest.
+  let n = 1
+  let oldEntry: RouterBFCacheEntry | null = prevActiveEntry
+  let clonedEntry: RouterBFCacheEntry = newActiveEntry
+  while (oldEntry !== null) {
+    if (oldEntry.stateKey === activeStateKey) {
+      // Fast path. This entry in the old list that corresponds to the key that
+      // is now active. We've already placed a clone of this entry at the front
+      // of the new list. We can reuse the rest of the old list without cloning.
+      // NOTE: We don't need to worry about eviction in this case because we
+      // haven't increased the size of the cache, and we assume the max size
+      // is constant across renders. If we were to change it to a dynamic limit,
+      // then the implementation would need to account for that.
+      clonedEntry.next = oldEntry.next
+      break
+    } else {
+      if (n >= MAX_BF_CACHE_ENTRIES) {
+        // We've reached the max number of entries. Delete the remaining entries
+        // from the cache.
+        let entry: RouterBFCacheEntry | null = oldEntry
+        do {
+          entry = entry.next
+        } while (entry !== null)
+        break
```

Comment:
you're not deleting anything here yet. but I assume you meant to detach each item in the list to avoid accidentally extending lifetimes or something otherwise you could just stop here and not even loop?

Reply:
oh lol I updated the PR to simplify some of the code (there used to be a map here that I needed to clean up) and I forgot to delete this part, good catch :D